pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go

```diff
@@ -536,11 +546,82 @@ func maybeCreateAndAddShardCol(
 	return shardColName
 }
 
-func createVirtualColumnForIndex(
-	b BuildCtx, tn *tree.TableName, tbl *scpb.Table, expr tree.Expr,
+func maybeCreateVirtualColumnForIndex(
+	b BuildCtx, tn *tree.TableName, tbl *scpb.Table, expr tree.Expr, inverted bool, lastColumn bool,
 ) string {
+	validateColumnIndexableType := func(t *types.T) {
+		if t.IsAmbiguous() {
+			panic(errors.WithHint(
+				pgerror.Newf(
+					pgcode.InvalidTableDefinition,
+					"type of index element %s is ambiguous",
+					expr.String(),
+				),
+				"consider adding a type cast to the expression",
+			))
+		}
+		// Check if the column type is indexable,
+		// non-inverted types.
+		if !inverted &&
+			!colinfo.ColumnTypeIsIndexable(t) {
+			panic(pgerror.Newf(
+				pgcode.InvalidTableDefinition,
+				"index element %s of type %s is not indexable",
+				expr,
+				t.Name()))
+		}
+		// Check if inverted columns are invertible.
+		if inverted &&
+			!lastColumn &&
+			!colinfo.ColumnTypeIsIndexable(t) {
+			panic(errors.WithHint(
+				pgerror.Newf(
+					pgcode.InvalidTableDefinition,
+					"index element %s of type %s is not allowed as a prefix column in an inverted index",
+					expr.String(),
+					t.Name(),
+				),
+				"see the documentation for more information about inverted indexes: "+docs.URL("inverted-indexes.html"),
+			))
+		}
+		if inverted &&
+			lastColumn &&
+			!colinfo.ColumnTypeIsInvertedIndexable(t) {
+			panic(errors.WithHint(
+				pgerror.Newf(
+					pgcode.InvalidTableDefinition,
+					"index element %s of type %s is not allowed as the last column in an inverted index",
+					expr,
+					t.Name(),
+				),
+				"see the documentation for more information about inverted indexes: "+docs.URL("inverted-indexes.html"),
+			))
+		}
+	}
 	elts := b.QueryByID(tbl.TableID)
-	colName := tabledesc.GenerateUniqueName("crdb_internal_idx_expr", func(name string) (found bool) {
+	colName := ""
+	// Check if any existing columns can satisfy this expression already.
```

Comment:
IIUC, you are looking for a column `e` (with ToPublic target) who has an compute expression that is equal to the compute expression of the to-be-added virtual column. If succeeded, we return early and hence re-using virtual column, right?

But where did you check this existing column `e` is also a virtual column? What if it's stored? Maybe we need the guardrails to be
```if target == scpb.ToPublic && e.ComputeExpr != nil && e.IsVirtual```

Reply:
Good point this should be e.IsVirtual && e.Inaccesible, so that user columns are never used.